<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>3-Data Curation • omicser</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="3-Data Curation">
<meta property="og:description" content="omicser">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">omicser</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.9.17092021</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/00_quickstart.html">0-Quickstart Overview</a>
    </li>
    <li>
      <a href="../articles/01_environment_setup.html">1-Environment Setup</a>
    </li>
    <li>
      <a href="../articles/02_install.html">2-Installation</a>
    </li>
    <li>
      <a href="../articles/03_data_curation.html">3-Data Curation</a>
    </li>
    <li>
      <a href="../articles/04_configuration.html">4-Database Configuration</a>
    </li>
    <li>
      <a href="../articles/05_browsing.html">5-Browsing</a>
    </li>
    <li>
      <a href="../articles/06_sharing.html">6-Sharing</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ndcn/omicser/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="03_data_curation_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>3-Data Curation</h1>
                        <h4 class="author">andy henrie</h4>
            
            <h4 class="date">9/14/2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ndcn/omicser/blob/master/vignettes/03_data_curation.Rmd"><code>vignettes/03_data_curation.Rmd</code></a></small>
      <div class="hidden name"><code>03_data_curation.Rmd</code></div>

    </div>

    
    
<p>The most crucial step is <em>curating</em> your data into a database that can be loaded by the browser. As the <em>curator</em> you will have the responsibility to make some choices about what and how the data can be seen.<br>
This process results in specifying the location of the data, and creating files which are formatted for the browser.</p>
<p>As an example there is the <a href="examples/curate_domenico_stem_cell.R"><code>curate_domenico_stem_cell.R</code></a> script in the /examples subdirectory, which illustrates the steps and some of the choices required to curate your data into a browsable <em>database</em>.</p>
<p>The process as illustrated in the example can be broken into the following steps:</p>
<ol style="list-style-type: decimal">
<li>
<em>provenance &amp; meta data setup</em> - define the meta-data and context for the dataset</li>
<li>
<em>raw data ingest</em> - translate the outputs of your QC to our database format 3a. <em>define helper function</em> - any helpers you need to read and process the outputs of your QC 3b.<em>load the raw data</em>
</li>
<li>
<em>pack into the browser data format</em> - pack into the anndata structure (scanpy/python)</li>
<li>
<em>post-processing</em> - compute relavent marginal quantities, define additional annotation and grouping variables, etc. 5a. <em>dimension reduction</em> - compute and cluster if needed</li>
<li>
<em>differential expression tables</em> - compute and/or formate existing tables</li>
<li>
<em>write database </em>- write the files to the database location</li>
</ol>
<div id="database-path" class="section level4">
<h4 class="hasAnchor">
<a href="#database-path" class="anchor"></a>Database Path</h4>
<p>The first step will be to make a folder to contain your data. We will call this folder and contents the <em>DATABASE</em> . Each <em>DATABASE</em> folder that one might want to load into the browser should be in the same path. e.g. the with the repositories <code>vignettes/</code> path there is a <code>test_db</code> folder which contains several sub-folder <em>DATABASES</em></p>
</div>
<div id="anndata-schema" class="section level4">
<h4 class="hasAnchor">
<a href="#anndata-schema" class="anchor"></a><code>AnnData</code> Schema</h4>
<p>The AnnData scheme requires us to define three pieces of data:</p>
<ol style="list-style-type: decimal">
<li>
<em>DATA</em> ($X): a matrix of measurements - e.g. transcriptomics - count matrix of cells by genes.</li>
<li>
<em><em>FEATURE</em> METADATA</em> ($var): a table of <code>omic</code> annotation - e.g. gene/protein names, families, “highly variable genes”, “is marker gene”, etc.</li>
<li>
<em><em>SAMPLE</em> METADATA</em> ($obs): a table of sample annotations - e.g. cell types, sample #, batch ID sex, experiemntal condition, etc.</li>
</ol>
<p>More info here <a href="https://cran.r-project.org/web/packages/anndata/readme/README.html" class="uri">https://cran.r-project.org/web/packages/anndata/readme/README.html</a></p>
<div class="figure">
<img src="../inst/anndata_for_r.png" alt=""><p class="caption">Anndata scheme</p>
</div>
</div>
<div id="loading-the-data" class="section level4">
<h4 class="hasAnchor">
<a href="#loading-the-data" class="anchor"></a>Loading the data</h4>
<p>In addition to these three pieces differential expression tables need to be pre-computed.</p>
<p>Here are some of the key helper functions and he section they fall into.</p>
<p>Here is an example of loading three data files and then pakaging them with a “helper function” (defined in section #2 of the example curation script) into the <code>data_list</code> which will be used by the next stage.</p>
</div>
<div id="setup_database" class="section level4">
<h4 class="hasAnchor">
<a href="#setup_database" class="anchor"></a><code>setup_database()</code>
</h4>
<p>The <code><a href="../reference/setup_database.html">omicser::setup_database()</a></code> function packages the separate tables – DATA matrix, omic FEATURE METADATA annotations, and SAMPLE METADATA – into the anndata object. This function can also take the name of a seurat object file.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu">omicser</span><span class="fu">::</span><span class="fu"><a href="../reference/setup_database.html">setup_database</a></span><span class="op">(</span>database_name<span class="op">=</span><span class="va">DB_NAME</span>,
                              data_in<span class="op">=</span><span class="va">data_list</span>,
                              db_meta<span class="op">=</span><span class="cn">NULL</span> ,
                              re_pack<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Once we have packed the data into the <code>anndata</code> object we can leverage <code>scanpy</code> and the <code>reticulate</code> python backend to do dimension reduction and clustering. Although this stage is not nescessary, it demonstrates the bridge to <code>cellxgene</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#==== 5-a. dimension reduction - PCA / umap ========================================================</span>
<span class="va">sc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/import.html">import</a></span><span class="op">(</span><span class="st">"scanpy"</span><span class="op">)</span>
<span class="co"># scanpy pre-processing - sc$pp</span>
<span class="va">sc</span><span class="op">$</span><span class="va">pp</span><span class="op">$</span><span class="fu">pca</span><span class="op">(</span><span class="va">adata</span><span class="op">)</span>
<span class="va">sc</span><span class="op">$</span><span class="va">pp</span><span class="op">$</span><span class="fu">neighbors</span><span class="op">(</span><span class="va">adata</span><span class="op">)</span>
<span class="va">sc</span><span class="op">$</span><span class="va">tl</span><span class="op">$</span><span class="fu">leiden</span><span class="op">(</span><span class="va">adata</span><span class="op">)</span>
<span class="va">sc</span><span class="op">$</span><span class="va">tl</span><span class="op">$</span><span class="fu">umap</span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></code></pre></div>
</div>
<div id="differential-expression-tables" class="section level4">
<h4 class="hasAnchor">
<a href="#differential-expression-tables" class="anchor"></a>Differential Expression Tables</h4>
<p>This is probably the trickiest part of the curation Fortunately we have some helper functions to help us compute them. Often – especially for <em>prote-</em> omic databases – differential expressions are computed as the output of the instrumentataion by a commercial software. These algorithms leverage bespoke statistics, so it will be best to reformat those tables.</p>
<div id="de-table-schema" class="section level5">
<h5 class="hasAnchor">
<a href="#de-table-schema" class="anchor"></a>DE Table Schema</h5>
<p>Most proteomic, metabelomic and lipidomic data will have differential calculations at the output of the instrumentation (which leverages know statastical assumptions of the quantifications) we can also use scanpys tools to compute differential expression. The diff_exp tables will be needed for volcano plots either way.</p>
<p>The differential expression table has these fields:</p>
<ul>
<li>
<code>group</code> - the comparison {names}V{reference}</li>
<li>
<code>names</code> - what are we comparing?</li>
<li>
<code>obs_name</code> - name of the meta data variable</li>
<li>
<code>test_type</code> - what statistic are we using</li>
<li>
<code>reference</code> - the denomenator. or the condition we are comparing expressions values to</li>
<li>
<code>comp_type</code> - grpVref or grpVrest. rest is all other conditions</li>
<li>
<code>logfoldchanges</code> - log2(name/reference)</li>
<li>
<code>scores</code> - statistic score</li>
<li>
<code>pvals</code> - pvalues from the stats test. e.g. t-test</li>
<li>
<code>pvals_adj</code> - adjusted pvalue (Q)</li>
<li>
<code>versus</code> - label which we will choose in the browser</li>
</ul>
</div>
<div id="omicsercompute_de_table" class="section level5">
<h5 class="hasAnchor">
<a href="#omicsercompute_de_table" class="anchor"></a>omicser::compute_de_table()</h5>
<p>In <a href="../R/pre_process_helpers.R"><code>R/pre_process_helpers.R</code></a> theres a function which leverages <code>scanpy</code> and the <code>anndata</code> format we have packed to do differential expression. We just need to pass a few quantites and it returns a properly formatted differential expression table.</p>
<p>parameters:</p>
<ul>
<li>
<code>adata</code> - the anndata object</li>
<li>
<code>comp_types</code> - what kind of comparisons? there are two types
<ul>
<li>“allVrest” which takes each of our experimental conditions in turn and compares against the “rest” of the data.</li>
<li>“{a}V{b}” or “firstgroupVsecondgroup” which compares the experimental condition “firstgroup” against “secondgroup”</li>
</ul>
</li>
<li>
<code>test_types</code> - statistical tests. See the examples or <code>scanpy</code> documentation for which test types are available.</li>
<li>
<code>obs_names</code> -name of the adata$obs column defining the comparision groups</li>
<li>
<code>sc</code> - the scanpy data object we imported with <code>reticulate</code>
</li>
</ul>
<p>Here’s an example which computes a differential expression with a <code>wilcoxon</code> test of significance for each <em>disease</em> with respect to the rest of the distease states, AND for each <em>cell_type</em> with-respect-to the “rest of” the <em>cell_type</em> s.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc</span> <span class="op">&lt;-</span> <span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/import.html">import</a></span><span class="op">(</span><span class="st">"scanpy"</span><span class="op">)</span>
<span class="va">test_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'wilcoxon'</span><span class="op">)</span>
<span class="va">comp_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'allVrest'</span><span class="op">)</span>
<span class="va">obs_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'disease'</span>,<span class="st">'cell_type'</span><span class="op">)</span>
<span class="va">diff_exp</span> <span class="op">&lt;-</span> <span class="fu">omicser</span><span class="fu">::</span><span class="fu"><a href="../reference/compute_de_table.html">compute_de_table</a></span><span class="op">(</span><span class="va">adata</span>,<span class="va">comp_types</span>, <span class="va">test_types</span>, <span class="va">obs_names</span>,<span class="va">sc</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="save-the-data-to-the-database" class="section level4">
<h4 class="hasAnchor">
<a href="#save-the-data-to-the-database" class="anchor"></a>Save the data to the <em>DATABASE</em>
</h4>
<p>Finally we write this the anndata data object to our <em>database_DATABASE</em> folder. In the examples contained in the <code>vignettes/</code> folder we defined <code>DS_ROOT_PATH &lt;- "test_db"</code>, and <code>DB_NAME &lt;- "domenico_stem_cell"</code> .</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">adata</span><span class="op">$</span><span class="fu">write_h5ad</span><span class="op">(</span>filename<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">DS_ROOT_PATH</span>,<span class="va">DB_NAME</span>,<span class="st">"db_data.h5ad"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>In the end each DATABASE folder should now these three files: 1. <code>db_data.h5ad</code> - the anndata object<br>
2. <code>db_de_table.rds</code> - differntial expression table 3, <code>db_meta.yml</code> - list of database ‘meta’ information</p>
<p>You might also want to save some <em>intermediate</em> files such as in the <a href="examples/curate_pbmc3k.R">examples</a> which also generate: <code>normalized_data.h5ad</code>,<code>core_data.h5ad</code>,and <code>norm_data_plus_dr.h5ad</code>.</p>
<p>We are almost there. In the next section we will cover the configuration which will add a <code>db_config.yml</code> files to the <em>DATABASE</em> directory, and create an <a href="omicser_options.yml"><code>omicser_options.yml</code></a> in the directory where the app will be executed.</p>
<p>On to <a href="04_configuration.html">CONFIGURATION</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Andy Henrie.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
